nomad/state/events_state_store.go

```diff
@@ -0,0 +1,53 @@
+package state
+
+import (
+	"fmt"
+
+	memdb "github.com/hashicorp/go-memdb"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// addNodeEvent is a function which wraps upsertNodeEvent
+func (s *StateStore) AddNodeEvent(index uint64, node *structs.Node, events []*structs.NodeEvent) error {
+	txn := s.db.Txn(true)
+	defer txn.Abort()
+
+	err := s.upsertNodeEvents(index, node, events, txn)
+	txn.Commit()
+	return err
+}
+
+// upsertNodeEvent upserts a node event for a respective node. It also maintains
+// that only 10 node events are ever stored simultaneously, deleting older
+// events once this bound has been reached.
+func (s *StateStore) upsertNodeEvents(index uint64, node *structs.Node, events []*structs.NodeEvent, txn *memdb.Txn) error {
+
+	// Copy the existing node
+	copyNode := new(structs.Node)
+	*copyNode = *node
+
+	nodeEvents := node.NodeEvents
+
+	for _, e := range events {
+		e.CreateIndex = index
+		e.ModifyIndex = index
+
+		// keep node events pruned to below 10 simultaneously
+		if len(nodeEvents) >= 10 {
```

Comment:
make this 10 a constant named something like max_retained_node_events and use it here and in the tests.